neuralop/layers/gno_block.py

```diff
@@ -0,0 +1,202 @@
+    channel_mlp_non_linearity : torch.nn function, optional
+        activation function for ChannelMLPLinear above, by default F.gelu
+    transform_type : str, optional
+        Which integral transform to compute. The mapping is:
+        'linear_kernelonly' -> (a)
+        'linear' -> (b) [DEFAULT]
+        'nonlinear_kernelonly' -> (c)
+        'nonlinear' -> (d)
+        If the input f is not given then (a) is computed
+        by default independently of this parameter.
+    use_open3d_neighbor_search: bool, optional
+        whether to use open3d or native-PyTorch search
+    use_torch_scatter_reduce : bool, optional
+        whether to reduce in integral computation using a function
+        provided by the extra dependency torch_scatter or the slower
+        native PyTorch implementation, by default True
+
+    References
+    -----------
+    [1]_ Neural Operator: Graph Kernel Network for Partial Differential Equations.
+        Zongyi Li, Kamyar Azizzadenesheli, Burigede Liu, Kaushik Bhattacharya, 
+        Anima Anandkumar. ArXiV, 2020 
+    """
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 coord_dim: int,
+                 radius: float,
+                 pos_embedding: nn.Module=None,
+                 channel_mlp: nn.Module=None,
+                 channel_mlp_layers: List[int]=None,
+                 channel_mlp_non_linearity=F.gelu,
+                 transform_type="linear",
+                 use_open3d_neighbor_search: bool=True,
+                 use_torch_scatter_reduce=True,):
+        super().__init__()
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.coord_dim = coord_dim
+
+        self.radius = radius
+
+        # Apply sinusoidal positional embedding
+        self.pos_embedding = pos_embedding
+                    
+        # Create in-to-out nb search module
+        if use_open3d_neighbor_search:
+            assert self.coord_dim == 3, f"Error: open3d is only designed for 3d data, \
+                GNO instantiated for dim={coord_dim}"
+        self.neighbor_search = NeighborSearch(use_open3d=use_open3d_neighbor_search)
+
+        # create proper kernel input channel dim
+        # if nonlinear of either type, add in_features dim
+        # otherwise just add x and y dim
+       
+        kernel_in_dim = self.coord_dim * 2
+        kernel_in_dim_str = "dim(y) + dim(x)"
+
+         # x and y dim will be embedding dim if pos embedding is applied
+        if self.pos_embedding is not None:
+            # Sinusoidal pos_embedding channels are multiplicative
+            if isinstance(self.pos_embedding, SinusoidalEmbedding2D):
+                kernel_in_dim *= self.pos_embedding.out_channels
+            # Grid pos_embedding channels are added to the end of inputs
+            elif isinstance(self.pos_embedding, GridEmbedding2D) or isinstance(self.pos, GridEmbeddingND):
```

Comment:
I guess we should raise a `ValueError` if `self.pos_embedding` is an instance of another class?

Reply:
Good idea.